core/shared/src/main/scala/scalaz/zio/Retry.scala

```diff
@@ -0,0 +1,381 @@
+// Copyright (C) 2018 John A. De Goes. All rights reserved.
+package scalaz.zio
+
+import scala.concurrent.duration.Duration
+import java.util.concurrent.TimeUnit
+
+/**
+ * A stateful strategy for retrying `IO` actions. See `IO.retryWith`.
+ */
+trait Retry[E, +S] { self =>
+
+  /**
+   * The full type of state used by the retry strategy, including hidden state
+   * not exposed via the `S` type parameter.
+   */
+  type State
+
+  /**
+   * Projects out the visible part of the state `S`.
+   */
+  def proj(state: State): S
+
+  /**
+   * The initial state of the strategy. This can be an effect, such as
+   * `nanoTime`.
+   */
+  val initial: IO[E, State]
+
+  /**
+   * Invoked on an error. This method can return the next state, which will continue
+   * the retry process, or it can return a failure, which will terminate the retry.
+   */
+  def update(e: E, s: State): IO[E, State]
+
+  /**
+   * Negates this strategy, returning failures for successes, and successes
+   * for failures.
+   */
+  def unary_! : Retry[E, S] = new Retry[E, S] {
+    type State = self.State
+
+    val initial = self.initial
+
+    def proj(state: State): S = self.proj(state)
+
+    def update(e: E, s: State): IO[E, State] =
+      self.update(e, s).redeem(_ => IO.now(s), _ => IO.fail(e))
+  }
+
+  /**
+   * Peeks at the visible part of the state, executes some action, and then
+   * continues retrying or not based on the specified predicate.
+   */
+  final def check[A](action: (E, S) => IO[E, A])(pred: A => Boolean): Retry[E, S] =
+    new Retry[E, S] {
+      type State = self.State
+
+      val initial = self.initial
+
+      def proj(state: State): S = self.proj(state)
+
+      def update(e: E, s: State): IO[E, State] =
+        for {
+          s <- self.update(e, s)
+          a <- action(e, proj(s))
+          _ <- if (pred(a)) IO.now(s) else IO.fail(e)
+        } yield s
+    }
+
+  /**
+   * Returns a new strategy that retries while the error matches the condition.
+   */
+  final def whileError(p: E => Boolean): Retry[E, S] =
+    check[E]((e, _) => IO.now(e))(p)
+
+  /**
+   * Returns a new strategy that retries until the error matches the condition.
+   */
+  final def untilError(p: E => Boolean): Retry[E, S] = !whileError(p)
+
+  /*
+   * Returns a new strategy that retries until the state matches the condition.
+   */
+  final def untilState(p: S => Boolean): Retry[E, S] = check[S]((_, s) => IO.now(s))(p)
+
+  /*
+   * Returns a new strategy that retries while the state matches the condition.
+   */
+  final def whileState(p: S => Boolean): Retry[E, S] = !untilState(p)
+
+  /**
+   * Returns a new strategy that retries for as long as this strategy and the
+   * specified strategy both agree to retry. For pure strategies (which have
+   * deterministic initial states/updates), the following law holds:
+   * {{{
+   * io.retryWith(r && r) === io.retryWith(r)
+   * }}}
+   */
+  final def &&[S2](that0: => Retry[E, S2]): Retry[E, (S, S2)] =
```

Comment:
This could have a named method too, like `par`. The combination of `par + map` could be called `parWith`.

Reply:
my 2 bikeshedding cents:
`par` suggests independence
> ... retries for as long as this strategy and the specified strategy both agree to retry...

this behaves more like "intersection" of 2 strategies, so `intersect` and `intersectWith` maybe? :)

Reply:
Are we still targeting programmers? I think at this point everybody is more than familiarized with `&&` and `||`

Reply:
We have to keep the operators. But some people prefer names, so usually I have a named method for every operator. For example, I have `andThen` for `<||>`.